pandas/_libs/algos.pyx

```diff
@@ -307,26 +308,44 @@ def nancorr_spearman(const float64_t[:, :] mat, Py_ssize_t minp=1):
     result = np.empty((K, K), dtype=np.float64)
     mask = np.isfinite(mat).view(np.uint8)
 
+    ranked_mat = np.empty((N, K), dtype=np.float64)
+    cached_ranks = set({})
```

Comment:
IIUC the whole caching thing you have here is to avoid the memory overhead of ranking columns up front right? If so I *think* it's a premature optimization so would probably just prefer if everything was ranked and we didn't do caching.

Reply:
So this approach would accept a hit on memory by preallocating `ranked_mat` (even this is probably not necessary), but save time by not calculating ranks up front that we end up not reusing.  Instead we wait until we need a full column set of ranks for _some_ calculation before computing and saving them.

Reply:
How much of a time savings is that? In any case your PR should offer a significant speed up by not ranking in each loop so I think this *might* still be OK.

Reply:
It was actually fairly negligible based on some local testing I did (maybe 5% slowdown), but it's probably worthwhile to add some more cases to `stat_ops.py` (wide tables, possibly a worst case missing pattern for this change, etc.)

Reply:
Did you see any perf difference here? Again would rather simplify and not have this if there's no major perf impact

Reply:
I'm actually seeing essentially the same performance from the previous implementation (ranking everything up front) on the "wide with nans" benchmark, so maybe just revert to that approach since the code is a lot easier to understand?